sdk/php/runtime/main.go

```diff
@@ -0,0 +1,171 @@
+) *PhpSdk {
+	if sdkSourceDir == nil {
+		// TODO: change back to git or something remote when the runtime is ready.
+		// Go back to ../ to get the root of the PHP SDK
+		// This is a bit of a hack, you need to call it from a parent directory but it works.
+		// Example from ../../ (the directory above the dagger git repo)
+		// dagger init --sdk=dagger/sdk/php --source=test
+		sdkSourceDir = dag.
+			CurrentModule().
+			Source().
+			Directory("../").
+			WithoutDirectory("runtime")
+	}
+
+	return &PhpSdk{
+		SourceDir:     sdkSourceDir,
+		RequiredPaths: []string{},
+		Container:     dag.Container().From(DefaultImage),
+	}
+}
+
+func (sdk *PhpSdk) Codegen(
+	ctx context.Context,
+	modSource *dagger.ModuleSource,
+	introspectionJSON *dagger.File,
+) (*dagger.GeneratedCode, error) {
+	ctr, err := sdk.CodegenBase(ctx, modSource, introspectionJSON)
+	if err != nil {
+		return nil, err
+	}
+
+	return dag.
+		GeneratedCode(ctr.Directory(ModSourceDirPath)).
+		WithVCSGeneratedPaths([]string{GenDir + "/**"}).
+		WithVCSIgnoredPaths([]string{GenDir}),
+		nil
+}
+
+func (sdk *PhpSdk) CodegenBase(
+	ctx context.Context,
+	modSource *dagger.ModuleSource,
+	introspectionJSON *dagger.File,
+) (*dagger.Container, error) {
+	ctr := sdk.Container
+
+	name, err := modSource.ModuleOriginalName(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("could not load module config: %w", err)
+	}
+
+	subPath, err := modSource.SourceSubpath(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("could not load module config: %w", err)
+	}
+
+	/**
+	 * Mounts the PHP SDK code,
+	 * Installs composer into the container
+	 * Runs composer install in the codegen directory
+	 * Runs codegen using the schema json provided by the dagger engine
+	 */
+	ctr = ctr.
+		WithMountedDirectory("/codegen", sdk.SourceDir).
+		WithoutEntrypoint().
+		WithWorkdir("/codegen").
+		WithExec([]string{"apk", "add", "git", "openssh", "curl"}).
+		WithExec([]string{
+			"git",
+			"config",
+			"--global",
+			"url.https://github.com/.insteadOf",
+			"git@github.com:",
+		}).
+		WithExec([]string{"./install-composer.sh"}).
+		WithExec([]string{"php", "composer.phar", "install"}).
+		WithMountedFile("schema.json", introspectionJSON).
+		WithExec([]string{
+			"./codegen",
+			"dagger:codegen",
+			"--schema-file",
+			"schema.json",
+		})
+
+	/**
+	 * Mounts the directory for the module we are generating for
+	 * Copies the generated code and rest of the sdk into the module directory under the sdk path
+	 * Runs the init template script for initialising a new module (this is a no-op if a composer.json already exists)
+	 */
+	ctr = ctr.
+		WithMountedDirectory(ModSourceDirPath, modSource.ContextDirectory()).
+		WithWorkdir(filepath.Join(ModSourceDirPath, subPath)).
+		WithDirectory(
+			GenDir,
+			ctr.Directory("/codegen"),
+			dagger.ContainerWithDirectoryOpts{Exclude: []string{
+				"codegen",
+				"runtime",
+				"vendor",
+				"install-composer.sh",
+				"composer.phar",
+				"composer.lock",
```

Comment:
It's ok to require the user to do a `dagger develop` first before running `composer install` locally. This happens for all SDKs.

Reply:
> [...] which won't have a fixed reference in the lock.

Wait, what does this mean? How does it show up in the lock?

Reply:
We've noticed that @carnage and I had different references to dagger in the composer.lock.

We're not entirely sure how composer is forming the reference for dagger since it installs dagger from generated code from a file path instead of a git repository.

Reply:
Doesn't matter if the path is generated. It should be the same relative path for both of you. Can you experiment and confirm this difference in the lock?

Reply:
Sorry, today is a new day and looking at this comment the lock file in question looks like it's from the SDK. In that case it's correct to exclude. What shouldn't be excluded is when you create a new module, a new lock file should be created and added to the source dir for the module. I suppose `compose install` will do that already.